### PR TITLE
review

### PR DESCRIPTION
**Pull request by Andrekukshinov**

### CODE REVIEW

File: src/main/java/by/kukshinov/tree/appliction/data/DataException.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+    public DataException(String message, Throwable cause) {
+	   super(message, cause);
+    }
+
+    public DataException(Throwable cause) {
+	   super(cause);
+    }
+
+    public DataException(
```
Comment (FilippStankevichTR) [inline src/main/java/by/kukshinov/tree/appliction/data/DataException.java]: ?

File: src/main/java/by/kukshinov/tree/appliction/data/DataReader.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package by.kukshinov.tree.appliction.data;
+
+public interface DataReader {
+    String readData(String filePath) throws DataException;
```
Comment (FilippStankevichTR) [inline src/main/java/by/kukshinov/tree/appliction/data/DataReader.java]: read

File: src/main/java/by/kukshinov/tree/appliction/data/ExpressionRecognizer.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package by.kukshinov.tree.appliction.data;
+
+public class ExpressionRecognizer {
```
Comment (FilippStankevichTR) [inline src/main/java/by/kukshinov/tree/appliction/data/ExpressionRecognizer.java]: ?

File: src/main/java/by/kukshinov/tree/appliction/data/FileDataReader.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+public class FileDataReader implements DataReader {
+    private static final char NEW_LINE = '\n';
+
+    public String readData(String filePath) throws DataException {
+	   Path path = Paths.get(filePath);
+	   StringBuilder resultBuilder = new StringBuilder();
+	   try {
+		  List<String> fileLines = Files.readAllLines(path, StandardCharsets.UTF_8);
+		  fileLines.forEach(line -> {
```
Comment (FilippStankevichTR) [inline src/main/java/by/kukshinov/tree/appliction/data/FileDataReader.java]: try to avoid side effects

Comment (FilippStankevichTR) [reply src/main/java/by/kukshinov/tree/appliction/data/FileDataReader.java]: fileLines.stream().collect(Joining(""))

File: src/main/java/by/kukshinov/tree/appliction/data/factory/ChainBuilder.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import by.kukshinov.tree.appliction.data.parser.TextParser;
+
+public class ChainBuilder {
+
+    public Parser build() {
+	   ExpressionRecognizer recognizer = new ExpressionRecognizer();
+        ParserFactory sentenceFactory = new SentenceParserFactory(recognizer);
+	   Parser sentenceParser = sentenceFactory.createParser();
+
+	   ParserFactory paragraphFactory= new ParagraphParserFactory(sentenceParser);
```
Comment (FilippStankevichTR) [inline src/main/java/by/kukshinov/tree/appliction/data/factory/ChainBuilder.java]: no sense in factory

File: src/main/java/by/kukshinov/tree/appliction/logics/ComponentProcessor.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    public Component calculateExpressions(Component component) {
+	   List<Component> newBuiltComponents = new ArrayList<>();
+	   List<Component> children = component.getChildren();
+	   if (children.size() != 0) {
+		  for (Component child : children) {
+			 Component builtChild = calculateExpressions(child);
+			 newBuiltComponents.add(builtChild);
+		  }
+		  return new CompositeComponent(newBuiltComponents);
+	   } else {
```
Comment (Andrekukshinov) [inline src/main/java/by/kukshinov/tree/appliction/logics/ComponentProcessor.java]: remove else

File: src/test/java/by/kukshinov/tree/appliction/data/ComponentDataProvider.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+import by.kukshinov.tree.appliction.model.Component;
+import by.kukshinov.tree.appliction.model.CompositeComponent;
+import by.kukshinov.tree.appliction.model.LexemeComponent;
+import org.testng.annotations.DataProvider;
+
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.List;
+
+public class ComponentDataProvider {
```
Comment (Andrekukshinov) [inline src/test/java/by/kukshinov/tree/appliction/data/ComponentDataProvider.java]: wtf? remove this sh**

**Pull request merged**